packages/core/src/shared/awsClientBuilderV3.ts

```diff
@@ -0,0 +1,150 @@
+
+export type AwsClient = IClient<any, any, any>
+interface AwsConfigOptions {
+    credentials: AwsCredentialIdentityProvider
+    region: string | Provider<string>
+    customUserAgent: UserAgent
+    requestHandler: any
+    apiVersion: string
+    endpoint: string
+}
+export type AwsClientOptions = AwsConfigOptions
+
+export interface AWSClientBuilderV3 {
+    createAwsService<C extends AwsClient>(
+        type: new (o: AwsClientOptions) => C,
+        options?: Partial<AwsClientOptions>,
+        region?: string,
+        userAgent?: boolean,
+        settings?: DevSettings
+    ): Promise<C>
+}
+
+export class DefaultAWSClientBuilderV3 implements AWSClientBuilderV3 {
+    public constructor(private readonly context: AwsContext) {}
+
+    private getShim(): CredentialsShim {
+        const shim = this.context.credentialsShim
+        if (!shim) {
+            throw new Error('Toolkit is not logged-in.')
+        }
+        return shim
+    }
+
+    public async createAwsService<C extends AwsClient>(
+        type: new (o: AwsClientOptions) => C,
+        options?: Partial<AwsClientOptions>,
+        region?: string,
+        userAgent: boolean = true,
+        settings?: DevSettings
+    ): Promise<C> {
+        const shim = this.getShim()
+        const opt = (options ?? {}) as AwsClientOptions
+
+        if (!opt.region && region) {
+            opt.region = region
+        }
+
+        if (!opt.customUserAgent && userAgent) {
+            opt.customUserAgent = [[getUserAgent({ includePlatform: true, includeClientId: true }), extensionVersion]]
+        }
+        // TODO: add tests for refresh logic.
+        opt.credentials = async () => {
+            const creds = await shim.get()
+            if (creds.expiration && creds.expiration.getTime() < Date.now()) {
+                return shim.refresh()
+            }
+            return creds
+        }
+
+        const service = new type(opt)
+        // TODO: add middleware for logging, telemetry, endpoints.
+        service.middlewareStack.add(telemetryMiddleware, { step: 'deserialize' } as DeserializeHandlerOptions)
+        return service
+    }
+}
+
+export function getServiceId(context: { clientName?: string; commandName?: string }): string {
+    return context.clientName?.toLowerCase().replace(/client$/, '') ?? 'unknown-service'
+}
+
+/**
+ * Record request IDs to the current context, potentially overriding the field if
+ * multiple API calls are made in the same context. We only do failures as successes are generally uninteresting and noisy.
+ */
+export function recordErrorTelemetry(err: Error, serviceName?: string) {
+    interface RequestData {
+        requestId?: string
+        requestServiceType?: string
+    }
+
+    telemetry.record({
+        requestId: getRequestId(err),
+        requestServiceType: serviceName,
+    } satisfies RequestData as any)
+}
+
+export function omitIfPresent<T extends Record<string, unknown>>(obj: T, keys: string[]): T {
```

Comment:
is this something that makes sense to be in a util?

maybe the other places we omit data isn't the same though?

Reply:
I think moving it to a util is a good idea. Nothing about it is specific to client builder.